src/java.base/share/classes/java/lang/Module.java

```diff
@@ -256,25 +258,74 @@ public ModuleLayer getLayer() {
     /**
      * Update this module to allow access to restricted methods.
      */
-    Module implAddEnableNativeAccess() {
+    synchronized Module implAddEnableNativeAccess() {
         enableNativeAccess = true;
         return this;
     }
 
     /**
-     * Update all unnamed modules to allow access to restricted methods.
+     * Returns {@code true} if this module can access
+     * <a href="foreign/package-summary.html#restricted"><em>restricted</em></a> methods.
+     *
+     * @return {@code true} if this module can access <em>restricted</em> methods.
      */
-    static void implAddEnableNativeAccessAllUnnamed() {
-        ALL_UNNAMED_MODULE.enableNativeAccess = true;
+    @PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
+    public boolean isNativeAccessEnabled() {
+        Module target = enableNativeAccessHolder();
+        synchronized(target) {
```

Comment:
I'm not sure how much of a problem this is, but I guess since the module object is publicly accessible, it's theoretically possible that some user code is also locking on it.

For the record; alternative could be to add a private field to module with a dedicated lock Object for enabling native access. Though, I don't think that's needed at the moment.

Reply:
That's a thought that occurred to me as well. E.g. declare the variable w/o `volatile`, but then use either Unsafe, or VarHandle to access the variable in `volatile` mode when required. That said, note that this will not get rid of all the synchronization needs: we would still need to synchronize on a module in the native access check to make sure that we issue at most one warning. But perhaps we can have a private lock just for that.

Reply:
I think it would be possible to emit only one warning by using a compare and exchange.

```
boolean shouldEnable = !ModuleBootstrap.hasEnableNativeAccessFlag();
isNativeAccessEnabled = Unsafe.compareAndExchangeBoolean(target, OFFSET, false, shouldEnable);
// check again with the safely read flag
if (isNativeAccessEnabled) {
    // another thread beat us to it - nothing to do
    return;
} else if (!shouldEnable) {
    throw new IllegalCallerException("Illegal native access from: " + this);
} else {
    // emit watning
}
```

Reply:
To be clear. The current implementation looks correct to me as well. So I'm fine with going ahead as is.